src/RtpFecQueue.c

```diff
@@ -0,0 +1,291 @@
+#include "Limelight-internal.h"
+#include "RtpFecQueue.h"
+#include "rs.h"
+
+#define ushort(x) (unsigned short) (x % (UINT16_MAX+1))
+#define isBefore(x, y) ushort(x - y) > (UINT16_MAX/2)
+
+void RtpfInitializeQueue(PRTP_FEC_QUEUE queue) {
+    fec_init();
+    memset(queue, 0, sizeof(*queue));
+    queue->nextRtpSequenceNumber = UINT16_MAX;
+    
+    queue->currentFrameNumber = UINT16_MAX;
+}
+
+void RtpfCleanupQueue(PRTP_FEC_QUEUE queue) {
+    while (queue->bufferHead != NULL) {
+        PRTPFEC_QUEUE_ENTRY entry = queue->bufferHead;
+        queue->bufferHead = entry->next;
+        free(entry->packet);
+    }
+}
+
+// newEntry is contained within the packet buffer so we free the whole entry by freeing entry->packet
+static int queuePacket(PRTP_FEC_QUEUE queue, PRTPFEC_QUEUE_ENTRY newEntry, int head, PRTP_PACKET packet) {
+    PRTPFEC_QUEUE_ENTRY entry;
+
+    // Don't queue packets we're already ahead of
+    if (isBefore(packet->sequenceNumber, queue->nextRtpSequenceNumber)) {
+        return 0;
+    }
+
+    // Don't queue duplicates either
+    entry = queue->bufferHead;
+    while (entry != NULL) {
+        if (entry->packet->sequenceNumber == packet->sequenceNumber) {
+            return 0;
+        }
+
+        entry = entry->next;
+    }
+    
+    newEntry->packet = packet;
+    newEntry->queueTimeMs = PltGetMillis();
+    newEntry->prev = NULL;
+    newEntry->next = NULL;
+
+    if (queue->bufferHead == NULL) {
+        LC_ASSERT(queue->bufferSize == 0);
+        queue->bufferHead = queue->bufferTail = newEntry;
+    }
+    else if (head) {
+        LC_ASSERT(queue->bufferSize > 0);
+        PRTPFEC_QUEUE_ENTRY oldHead = queue->bufferHead;
+        newEntry->next = oldHead;
+        LC_ASSERT(oldHead->prev == NULL);
+        oldHead->prev = newEntry;
+        queue->bufferHead = newEntry;
+    }
+    else {
+        LC_ASSERT(queue->bufferSize > 0);
+        PRTPFEC_QUEUE_ENTRY oldTail = queue->bufferTail;
+        newEntry->prev = oldTail;
+        LC_ASSERT(oldTail->next == NULL);
+        oldTail->next = newEntry;
+        queue->bufferTail = newEntry;
+    }
+    queue->bufferSize++;
+
+    return 1;
+}
+
+static void queueRecoveredPacket(PRTP_FEC_QUEUE queue, PRTPFEC_QUEUE_ENTRY entry, PRTP_PACKET packet, int sequenceNumber) {
+    //Correct RTP header as it isn't in the FEC data
+    packet->header = queue->queueHead->packet->header;
+    packet->sequenceNumber = sequenceNumber;
+    
+    int dataOffset = sizeof(*packet);
+    if (packet->header & FLAG_EXTENSION) {
+       dataOffset += 4; // 2 additional fields
+    }
+
+    PNV_VIDEO_PACKET nvPacket = (PNV_VIDEO_PACKET)(((char*)packet) + dataOffset);
+    nvPacket->frameIndex = queue->currentFrameNumber;
+    
+    //Set size of generated packet
+    int size = StreamConfig.packetSize + dataOffset;
+    
+    //Remove padding if it's the last packet in frame 
```

Comment:
No, I already encountered some artifacts myself. So probably I just remove this.